publish-mqtt/src/homie.rs

```diff
@@ -0,0 +1,204 @@
+use async_channel::{SendError, Sender};
+use rumqttc::{self, EventLoop, LastWill, MqttOptions, Publish, QoS, Request};
+use std::error::Error;
+use tokio::task::{self, JoinHandle};
+
+const HOMIE_VERSION: &str = "4.0";
+
+#[derive(Clone, Debug)]
+pub struct Property {
+    id: String,
+    name: String,
+    datatype: String,
+    unit: String,
+}
+
+impl Property {
+    pub fn new(id: &str, name: &str, datatype: &str, unit: &str) -> Property {
+        Property {
+            id: id.to_string(),
+            name: name.to_string(),
+            datatype: datatype.to_string(),
+            unit: unit.to_string(),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Node {
+    id: String,
+    name: String,
+    node_type: String,
+    properties: Vec<Property>,
+}
+
+impl Node {
+    pub fn new(id: String, name: String, node_type: String, properties: Vec<Property>) -> Node {
+        Node {
+            id,
+            name,
+            node_type,
+            properties,
+        }
+    }
+}
+
+pub(crate) struct HomieDevice {
+    requests_tx: Sender<Request>,
+    device_base: String,
+    device_name: String,
+    nodes: Vec<Node>,
+}
+
+impl HomieDevice {
+    pub async fn new(
+        device_base: &str,
+        device_name: &str,
+        mut mqtt_options: MqttOptions,
+    ) -> (
+        HomieDevice,
+        JoinHandle<Result<(), Box<dyn Error + Send + Sync>>>,
+    ) {
```

Comment:
yeah, now that I look at it, ::new() returning something other than Self is super-surprising. In other languages I would probably make a free function `connect(device_base, device_name, options) -> (HomieDevice, JoinHandle)`

Not sure what the best practice for rust is. Maybe HomieDevice::spawn(...) => (HomieDevice, JoinHandle)

Reply:
Renamed to spawn.